server/polar/order/service.py

```diff
@@ -1282,7 +1296,84 @@ async def _handle_consecutive_dunning_attempts(
         self, session: AsyncSession, order: Order
     ) -> Order:
         """Handle consecutive dunning attempts for an order."""
-        # TODO : Implement logic for handling consecutive dunning attempts
+        failed_attempts = await self._count_failed_payment_attempts(session, order)
+
+        if failed_attempts >= len(settings.DUNNING_RETRY_INTERVALS):
+            # No more retries, mark subscription as unpaid and clear retry date
+            repository = OrderRepository.from_session(session)
+            order = await repository.update(
+                order, update_dict={"next_payment_attempt_at": None}
+            )
+
+            if order.subscription is not None:
+                await subscription_service.cancel(session, order.subscription)
+
+            return order
+
+        # Schedule next retry using the appropriate interval
+        next_interval = settings.DUNNING_RETRY_INTERVALS[failed_attempts]
+        next_retry_date = utc_now() + next_interval
+
+        repository = OrderRepository.from_session(session)
+        order = await repository.update(
+            order, update_dict={"next_payment_attempt_at": next_retry_date}
+        )
+
+        return order
+
+    async def _count_failed_payment_attempts(
+        self, session: AsyncSession, order: Order
+    ) -> int:
+        statement = select(func.count(Payment.id)).where(
```

Comment:
We have statements in the service, but we also have the repository. What is the logic behind it? 🤔